fuzzers/main.py

```diff
@@ -0,0 +1,819 @@
+            f'artifacts/{self.target["runner"]}/{artifact}')
+        client.send_message({
+            'type':
+                'stream',
+            'to':
+                'nearinc/fuzzer/private',
+            'topic':
+                f'{branch}: artifact {artifact}',
+            'content': (
+                f'# Fuzzer found new crash for branch *{branch}* {already_reported_msg}\n'
+                f'Full logs are available at {logs_url}.\n'
+                f'\n'
+                f'You can download the artifact by using the following command (all commands '
+                f'are to be run from the root of `nearcore`):\n'
+                f'```\n'
+                f'gsutil cp {gcs_artifact} {artifact_dest}\n'
+                f'```\n'
+                f'\n'
+                f'Then, you can reproduce by running the following command:\n'
+                f'```\n'
+                f'cd {quoted_crate}\n'
+                f'RUSTC_BOOTSTRAP=1 cargo fuzz run {quoted_runner} {quoted_artifact}\n'
+                f'```\n'
+                f'\n'
+                f'Or minimize by running the following command:\n'
+                f'```\n'
+                f'cd {quoted_crate}\n'
+                f'RUSTC_BOOTSTRAP=1 cargo fuzz tmin {quoted_runner} {quoted_artifact}\n'
+                f'```\n'
+                f'\n'
+                f'Please edit the topic name to add more meaningful information once investigated. '
+                f'Keeping the artifact hash in it can help if the same artifact gets detected '
+                f'as crashing another branch.\n'),
+        })
+        last_log_lines = ''
+        for line in log_lines[::-1]:
+            if line.startswith('```'):
+                # Censor the end of a spoiler block, not great but this is for human consumption
+                # anyway
+                line = ' ' + line
+            if len(last_log_lines) + len(line) > 9000:
+                # Zulip limit is 10k, let's keep some safety buffer here
+                break
+            last_log_lines = line + last_log_lines
+        client.send_message({
+            'type': 'stream',
+            'to': 'nearinc/fuzzer/private',
+            'topic': f'{branch}: artifact {artifact}',
+            'content': f'```spoiler Last few log lines\n{last_log_lines}\n```',
+        })
+
+    def signal(self, sig: int) -> None:
+        """
+        Signal the fuzzer process (with all its process group as a fuzzer can spawn sub-fuzzers)
+        """
+
+        print(f'Sending signal {sig} to fuzzer {self.proc.pid}')
+        os.killpg(os.getpgid(self.proc.pid), sig)
+
+
+# Actually this should also require a ['weight']: int bound but it seems like a mess to encode
+T = typing.TypeVar('T')  # pylint: disable=invalid-name
+
+
+def random_weighted(array: typing.List[T], num: int) -> typing.List[T]:
+    untyped_array = typing.cast(typing.Any, array)
+    return random.choices(array, [x['weight'] for x in untyped_array], k=num)
```

Comment:
```python
T = typing.TypeVar('T', tying.Dict[str, typing.Any])  # pylint: disable=invalid-name

def random_weighted(array: typing.List[T], num: int) -> typing.List[T]:
    return random.choices(array, [x['weight'] for x in array], k=num)
```

should work

Reply:
I'm getting `TypeVar cannot have only a single constraint`, and am not sure how to fix that without losing the information that random_weighted returns a list of the same type as the input

Reply:
Sorry, yes, it should be bound: `T = typing.TypeVar('T', bound=typing.Dict[str, typing.Any])`. You might try `bound=TypedDict('T', {'weight': int})` though I dunno if that would work.